File: src/main/java/com/raychenon/leetcode/MergeIntervals.java
package com.raychenon.leetcode;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * User: raychenon
 * Date: 2020-02-02
 * https://leetcode.com/problems/merge-intervals/
 */
public class MergeIntervals {

    /**
     * Time complexity: O(N * Log(N))  due to sorting
     * Space complexity: O(1)
     */
    public int[][] merge(int[][] intervals) {
        if (intervals.length <= 1)
            return intervals;

        // Sort by ascending the left integer
        Arrays.sort(intervals, (i1, i2) -> Integer.compare(i1[0], i2[0]));

        List<int[]> result = new LinkedList<>();
        int[] newInterval = intervals[0];
        result.add(newInterval);
        for (int[] interval : intervals) {
            if (interval[0] <= newInterval[1]) // Overlapping intervals, move the end if needed
                newInterval[1] = Math.max(newInterval[1], interval[1]);
            else {                             // Disjoint intervals, add the new interval to the list
                result.add(interval);
                newInterval = interval;
            }
        }

        return result.toArray(new int[result.size()][]);
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/MergeIntervals.kt
package com.raychenon.kotlin.leetcode

import java.util.Arrays

/**
 * User: raychenon
 * Date: 2020-01-30
 * https://leetcode.com/problems/merge-intervals/
 */
object MergeIntervals {

    /**
     * Time complexity: O(N * Log(N))
     * Space complexity: O(1)
     */
    fun merge(intervals: Array<IntArray>): Array<IntArray> {

        if (intervals.size <= 1) return intervals

        // Sort by ascending starting point
        // sorting is n*log(n) operation
        Arrays.sort(
            intervals
        ) { i1: IntArray, i2: IntArray ->
            Integer.compare(
                i1[0],
                i2[0]
            )
        }
        var arr = mutableListOf<IntArray>()
        var currInterval = intervals.get(0)
        arr.add(currInterval)

        for (interval in intervals) {
            // if the interval overlap, change the biggest element of the interval
            if (currInterval.get(1) >= interval.get(0)) {
                currInterval[1] = Math.max(currInterval.get(1), interval.get(1))
            } else {
                arr.add(interval)
                currInterval = interval
            }
        }

        return arr.toTypedArray()
    }
}
